ring/partition_instance_ring.go

```diff
@@ -63,46 +63,64 @@ func (r *PartitionInstanceRing) GetReplicationSetsForOperation(op Operation) ([]
 	zonesBuffer := make([]string, 0, 3) // Pre-allocate buffer assuming 3 zones.
 
 	for partitionID := range partitionsRingDesc.Partitions {
-		ownerIDs := partitionsRing.PartitionOwnerIDs(partitionID)
-		instances := make([]InstanceDesc, 0, len(ownerIDs))
-
-		for _, instanceID := range ownerIDs {
-			instance, err := r.instancesRing.GetInstance(instanceID)
-			if err != nil {
-				// If an instance doesn't exist in the instances ring we don't return an error
-				// but lookup for other instances of the partition.
-				continue
-			}
-
-			if !instance.IsHealthy(op, r.heartbeatTimeout, now) {
-				continue
-			}
-
-			instances = append(instances, instance)
+		replicationSet, err := r.getReplicationSetForPartitionAndOperation(partitionID, op, now, zonesBuffer)
+		if err != nil {
+			return nil, err
 		}
 
-		if len(instances) == 0 {
-			return nil, fmt.Errorf("partition %d: %w", partitionID, ErrTooManyUnhealthyInstances)
-		}
+		result = append(result, replicationSet)
+	}
+	return result, nil
+}
 
-		// Count the number of unique zones among instances.
-		zonesBuffer = uniqueZonesFromInstances(instances, zonesBuffer[:0])
-		uniqueZones := len(zonesBuffer)
+// GetReplicationSetForPartitionAndOperation returns a ReplicationSet for the input partition. If the partition doesn't
+// exist or there are no healthy owners for the partition, an error is returned.
+func (r *PartitionInstanceRing) GetReplicationSetForPartitionAndOperation(partitionID int32, op Operation) (ReplicationSet, error) {
+	zonesBuffer := make([]string, 0, 3) // Pre-allocate buffer assuming 3 zones.
```

Comment:
I think we can avoid the allocation here completely.

```suggestion
    var stackZonesBuffer [3]string // Pre-allocate buffer assuming 3 zones.
	zonesBuffer := stackZonesBuffer[:]

```

(I'd also do 5, not 3, let's not be greedy)

Reply:
Picked the on-stack buffer: https://github.com/grafana/dskit/pull/623/commits/bf3dc4a5014f77d2249630b712cbeee12e4c2bcf. About 3 vs 5 I will leave it for a separate discussion, given in the ring/ pkg we assume 3 in different places.